### ISSUE TITLE
Published to Rubygems.

### ISSUE THREAD

**Issue by a-bash** [grempe/secretsharing#1]
I'm very impressed with this implementation and looking forward to using it my own project. I notice that you have contributed some great code clean up to the originally published gem. 

Will these changes be contributed back to a new version of the published 'secretsharing' gem or will they remain on your fork? I'm just trying to figure out the best way to get your latest code into my project.

Thanks!
Andy

**Comment by grempe** [comment]
Hi Andy,

Thanks for the nice comments. I'll give most of the credit to the original implementation which I cleaned up for potential use in a project I've had in the back of my mind for a while, and also to help me better understand the original code.

There are no plans at the moment to contribute back to the original source (which has not been updated since before I forked it) or to publish as a gem.  Since its security sensitive it's something I would consider doing only after there has been more real world testing and review.  Perhaps you can use it in a Gemfile direct from this repo (in its current experimental state) and provide some pull requests or suggestions once you use it?

I'm curious what type of project you plan to use it in if that is something you can share.

Cheers,

Glenn

**Comment by grempe** [comment]
Please note that there is a version of the gem already published at:

http://rubygems.org/gems/secretsharing

Which is Alexander's old 0.3 version of the gem.  I think mine is in better shape organizationally and test-wise. I don't have access to that gem on rubygems.org so he'd either have to accept my changes or I would have to fork and rename the gem to something new.  Not sure which approach I would prefer at this point.  We'd have to test against his API and mine and determine if there are any breaking changes before doing so and potentially breaking someone's code that is depending on his gem.

Its all a little vague in my mind at the moment since I have not looked at this code in a while. :-)

Glenn

**Comment by a-bash** [comment]
Glenn,

Great. Thanks for clarifying. I agree that your code in better shape organizationally so I'm going to jump off from your fork. 

This algorithm is going to play a big role in my project going forward so I would expect to see pull requests coming back to you. I don't have a good suggestion yet on getting the latest up to Ruby Gems. I'll give it some more thought and get back to you.

As I see it, it's unlikely that any of us are going to improve on the fundamental algorithm so it really comes down to organizational clarity and test coverage. I'm still in exploring mode but I'm really happy so far with where the code is at.

This is what I'm working on:
http://www.vault-tree.org/
https://github.com/VaultTree/vault-tree
https://github.com/VaultTree/contracts

Andy

**Comment by grempe** [comment]
Andy, FYI I've pushed a bunch of commits to github today.  I don't think this will be backwards incompatible for you but there are a number of enhancements and some spit and polish applied.

I am also thinking that I will abstract the Secret into its own class so that retrieving a secret (after you've added all the shares) will give back a common type of object that you can then manipulate in different ways (e.g. get back a Base64 version of the secret, or a OpenSSL Big Num).

So keep an eye on the repo.  Once things settle down I'll reach out to Alex to see if he'll mind me taking over the existing gem on RubyGems.  If not I'll publish this fork under a new Gem name so it won't conflict.

Pull requests and feedback still welcome.

Glenn

**Comment by a-bash** [comment]
That's perfect. Thank you. If you do feel the need to break backwards compatibility you won't cause problems on my end. In my application I'll be handing around only the Base64 and Hex representations of the secrets and shares, not the objects. If there are breaking change I'll only have to fix it one or two places.

**Comment by grempe** [comment]
Yes, as I've gotten into it there are breaking changes, but I personally like the layout better.  I'm still working on it.  There will be a Container object which can be injected with Share objects to produce a Secret object.  The Secret has methods to get different representations out of it (e.g. to_s, to_base64).  I think the Base64 will need some work since right now it only goes in 1 direction (to_base64).  I think the Secret object should also accept a either a base64 representation in Secret.new(b64string) or Secret.from_base64(b64string) which would allow base64 to be a nice compact representation of the secret which at least for now is always an OpenSSL::BN object.

**Comment by grempe** [comment]
New commits pushed with additional refactoring of Container and Secret.  Should be reaching some stability for the API now.

**Comment by grempe** [comment]
New simple command line client pushed to let you both generate shares from a secret and retrieve a secret from shares.  I've reached out to Alexander Klink to see about him adding me as an owner on the http://rubygems.org account.

I think there are a couple of things I would like to finish adding before a public release.  Foremost among those is the ability to pass in an arbitrary UTF-8 text secret and generate shares from it (instead of being limited to an OpenSSL::BN as you are now).  Internally it would still be represented as an OpenSSL::BN.

I am also thinking about some changes to the format of the shares.  I think it might be useful to add the number of shares required to reconstitute the share to the share itself so the operator doesn't have to remember that.  I am also thinking about adding a SHA256 hash of the original secret to the share to allow verification that the original secret was in fact re-created.  Currently, for example, if you are required to have 3 shares (k value) to retrieve a secret you could be given two valid shares and one cheater share.  The calculations would still run and a secret would be output.  It just would not be the original secret and the operator would have no way of knowing that if the secret output was correct or not.  I think if the share included a hash of the original secret you would be able to not only visually see that the shares don't match, but you could teach the program to run the resulting secret through that same hash routine to see if there is a match and alert the operator if the secret retrieved was in fact the one originally stored.  I am not sure if there are security implications to this.  I think there would only be implications if SHA256 were to become reversible which I think is unlikely.

Once those changes are made, assuming Alexander is OK, then I think a 1.0 release would be forthcoming.

**Comment by grempe** [comment]
Alexander just announced on Twitter that he made me a co-owner of the gem on rubygems.org which is great news.

https://twitter.com/alech/status/410075813263187968

**Comment by a-bash** [comment]
That's good news. All your changes for version 1.0 sound great. I look forward
to the upgrade once it makes it up to RG. Feel free to close this issue.

**Comment by dexterp** [comment]
Hi Glen,
I was looking for a Shamirs Secret sharing module and ruby and found your fork which is just what I'm looking for. I got a little confused when I ran 'gem install secretsharing' only to realize that the module on RG is now completely different from your fork. I read from the threads your planning to put this gem into RG. I was wondering if you had already done so perhaps as a different name? I'm quite keen to use it.

**Comment by grempe** [comment]
Hello dexterp.  Sorry, I have not yet replaced the version on ruby gems with my fork.  I just had my first child and she's keeping me pretty busy from finishing up some things before I do so. :-)  In the interim you can install the gem using bundler and specifying the git repo, or download the repo and use the installation rake task.

Cheers,

Glenn

**Comment by a-bash** [comment]
Glenn,
Any outstanding work that I can help with to get the next version up on Ruby Gems. Certainly no feature requests from me, I'm just hoping to get the great clean up commits that you pushed up since you took over.

Thanks
Andy

**Comment by grempe** [comment]
FYI, version 1.0.0 of the gem has been released to rubygems. Please let me know if you see any issues.

https://rubygems.org/gems/secretsharing

:-)
